### ISSUE TITLE
Any way to reuse 2 HTTP keep-alive connections to the same server in parallel ?

### ISSUE THREAD

**Issue by wreuven** [warmcat/libwebsockets#2847]
If I specify LCCSCF_PIPELINE then my connection requests get serialized onto the same socket so resuse but only one GET at a time.
If I don't specify LCCSCF_PIPELINE, then LWS closes the connection (connection: close header) after receiving a response so no reuse.

The use-case is that I want to download audio and video in parallel from the same CDN and reuse both connections with keep-alive. Seems there should be some way to do this.

**Comment by ljluestc** [comment]
```

#include <libwebsockets.h>
#include <stdio.h>
#include <string.h>

struct per_session_data {
    // You can store any session-specific data here
    int download_type; // 0 for audio, 1 for video
};

// Callback function for handling HTTP connections
static int callback_http(struct lws *wsi, enum lws_callback_reasons reason,
                          void *user, void *in, size_t len) {
    struct per_session_data *pss = (struct per_session_data *)user;

    switch (reason) {
        case LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP:
            lwsl_info("HTTP client connection established.\n");
            break;

        case LWS_CALLBACK_COMPLETED_CLIENT_HTTP:
            lwsl_info("HTTP request completed for type %d.\n", pss->download_type);
            // Handle the received data
            break;

        case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
            lwsl_err("Client connection error occurred.\n");
            break;

        case LWS_CALLBACK_CLOSED_CLIENT_HTTP:
            lwsl_info("Client connection closed.\n");
            break;

        default:
            break;
    }
    return 0;
}

// Function to initiate downloads
void initiate_download(struct lws_context *context, const char *url, int download_type) {
    struct per_session_data *pss = malloc(sizeof(struct per_session_data));
    pss->download_type = download_type;

    // Set up the connection info
    struct lws_client_connect_info client_info = {0};
    client_info.context = context;
    client_info.address = "example.com"; // Replace with your CDN
    client_info.port = 80; // Use 443 for HTTPS
    client_info.path = url; // Path to the audio/video resource
    client_info.host = client_info.address;
    client_info.origin = client_info.address;
    client_info.protocol = "http-client-protocol"; // Ensure your protocol is set
    client_info.userdata = pss; // Pass session data

    // Initiate the connection
    lws_client_connect_via_info(&client_info);
}

int main(void) {
    struct lws_context_creation_info info;
    struct lws_context *context;

    memset(&info, 0, sizeof(info));
    info.port = CONTEXT_PORT_NO_LISTEN; // Client context
    info.protocols = (struct lws_protocols[]) {
        {
            "http-client-protocol",
            callback_http,
            sizeof(struct per_session_data),
            0,
        },
        { NULL, NULL, 0, 0 } // terminator
    };

    context = lws_create_context(&info);
    if (!context) {
        lwsl_err("Failed to create LWS context\n");
        return -1;
    }

    // Start the downloads
    initiate_download(context, "/path/to/audio", 0); // Download audio
    initiate_download(context, "/path/to/video", 1); // Download video

    // Main event loop
    while (1) {
        lws_service(context, 0);
    }

    lws_context_destroy(context);
    return 0;
}
```

**Comment by lws-team** [comment]
@ljluestc thanks for your efforts making these related examples.

**Comment by wreuven** [comment]
> ```
> 
> #include <libwebsockets.h>
> #include <stdio.h>
> #include <string.h>
> 
> struct per_session_data {
>     // You can store any session-specific data here
>     int download_type; // 0 for audio, 1 for video
> };
> 
> // Callback function for handling HTTP connections
> static int callback_http(struct lws *wsi, enum lws_callback_reasons reason,
>                           void *user, void *in, size_t len) {
>     struct per_session_data *pss = (struct per_session_data *)user;
> 
>     switch (reason) {
>         case LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP:
>             lwsl_info("HTTP client connection established.\n");
>             break;
> 
>         case LWS_CALLBACK_COMPLETED_CLIENT_HTTP:
>             lwsl_info("HTTP request completed for type %d.\n", pss->download_type);
>             // Handle the received data
>             break;
> 
>         case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
>             lwsl_err("Client connection error occurred.\n");
>             break;
> 
>         case LWS_CALLBACK_CLOSED_CLIENT_HTTP:
>             lwsl_info("Client connection closed.\n");
>             break;
> 
>         default:
>             break;
>     }
>     return 0;
> }
> 
> // Function to initiate downloads
> void initiate_download(struct lws_context *context, const char *url, int download_type) {
>     struct per_session_data *pss = malloc(sizeof(struct per_session_data));
>     pss->download_type = download_type;
> 
>     // Set up the connection info
>     struct lws_client_connect_info client_info = {0};
>     client_info.context = context;
>     client_info.address = "example.com"; // Replace with your CDN
>     client_info.port = 80; // Use 443 for HTTPS
>     client_info.path = url; // Path to the audio/video resource
>     client_info.host = client_info.address;
>     client_info.origin = client_info.address;
>     client_info.protocol = "http-client-protocol"; // Ensure your protocol is set
>     client_info.userdata = pss; // Pass session data
> 
>     // Initiate the connection
>     lws_client_connect_via_info(&client_info);
> }
> 
> int main(void) {
>     struct lws_context_creation_info info;
>     struct lws_context *context;
> 
>     memset(&info, 0, sizeof(info));
>     info.port = CONTEXT_PORT_NO_LISTEN; // Client context
>     info.protocols = (struct lws_protocols[]) {
>         {
>             "http-client-protocol",
>             callback_http,
>             sizeof(struct per_session_data),
>             0,
>         },
>         { NULL, NULL, 0, 0 } // terminator
>     };
> 
>     context = lws_create_context(&info);
>     if (!context) {
>         lwsl_err("Failed to create LWS context\n");
>         return -1;
>     }
> 
>     // Start the downloads
>     initiate_download(context, "/path/to/audio", 0); // Download audio
>     initiate_download(context, "/path/to/video", 1); // Download video
> 
>     // Main event loop
>     while (1) {
>         lws_service(context, 0);
>     }
> 
>     lws_context_destroy(context);
>     return 0;
> }
> ```

Thanks for this. I ended up creating a different VHOST to solve this, but this looks simple.
Note that is seems to me that the example code is leaking the pss memory on each download.

**Comment by lws-team** [comment]
> leak

Yes it needs to free the manually-prepared user data when either `LWS_CALLBACK_CLIENT_CONNECTION_ERROR` (the client connection never connected and failed) or `LWS_CALLBACK_CLOSED_CLIENT_HTTP` (transaction's done) happens.
